archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,872 @@
+  intro n,
+  calc
+  log_stirling_seq 1 - log_stirling_seq n.succ = log_stirling_seq' 0 - log_stirling_seq' n : rfl
+    ... = ∑ k in range n, (log_stirling_seq' k - log_stirling_seq' (k + 1)) :
+    by rw ← (sum_range_sub' log_stirling_seq' n)
+    ... = ∑ k in range n, (log_stirling_seq k.succ - log_stirling_seq k.succ.succ) : rfl
+    ... ≤ ∑ k in range n, 1 / (4 * k.succ * k.succ.succ) :
+    sum_le_sum (λ k, λ hk, log_stirling_seq_sub_log_stirling_seq_succ k)
+    ... = ∑ k in range n, (1 / 4) * (1 / (k.succ * k.succ.succ)) :
+    begin
+      have hi : ∀ (k : ℕ), (1 : ℝ) / (4 * k.succ * k.succ.succ) =
+      (1 / 4) * (1 / (k.succ * k.succ.succ)) :=
+      begin
+        intro k,
+        norm_cast,
+        field_simp,
+        simp only [one_div, inv_inj],
+        ring_nf,
+      end,
+      refine sum_congr rfl (λ k, λ hk, hi k),
+    end
+    ... = 1 / 4 * ∑ k in range n, 1 / (k.succ * k.succ.succ) : by rw mul_sum
+    ... ≤ 1 / 4 * 1 :
+    begin
+      refine (mul_le_mul_left _).mpr _,
+      { exact div_pos one_pos four_pos },
+      { convert rat.cast_le.mpr (partial_sum_consecutive_reciprocals n),
+        rw rat_cast_sum,
+        push_cast,
+        exact rat.cast_one.symm },
+    end
+    ... = 1 / 4 : by rw mul_one,
+end
+
+/-- The sequence `log_stirling_seq` is bounded below by `3/4 - 1/2 * log 2`  for `n ≥ 1`. -/
+lemma log_stirling_seq_bounded_by_constant : ∀ (n : ℕ), 3 / (4 : ℝ) - 1 / 2 * log 2
+≤ log_stirling_seq n.succ :=
+begin
+  intro n,
+  calc
+  log_stirling_seq n.succ ≥ log_stirling_seq 1 - 1 / 4 : sub_le.mp (log_stirling_seq_bounded_aux n)
+    ... = (log ((1 : ℕ).factorial) - 1 / 2 * log (2 * (1 : ℕ)) - (1 : ℕ) *
+          log ((1 : ℕ) / (exp 1))) - 1 / 4 : by rw log_stirling_seq_formula 0
+    ... = 0 - 1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 :
+    by simp only [factorial_one, cast_one, log_one, one_div, mul_one, log_inv, log_exp, mul_neg]
+    ... = -1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 : by ring
+    ... = -1 / 2 * log 2 + 1 - 1 / 4  : by simp only [one_div, log_inv, log_exp, sub_neg_eq_add]
+    ... = -1 / 2 * log 2 + 3 / 4      : by ring
+    ... = 3 / (4 : ℝ) - 1 / 2 * log 2 : by ring,
+end
+
+/--
+Any sequence `b` of real numbers that is monotone decreasing and bounded below has
+a limit in the real numbers.
+-/
+lemma monotone_convergence (b : ℕ → ℝ) (h_sd : ∀ (n m : ℕ), n ≤ m → b m ≤ b n)
+  (h_bounded : (lower_bounds (set.range b)).nonempty) : ∃ (m : ℝ), tendsto b at_top (𝓝 m) :=
+begin
+  use Inf (set.range b),
+  exact tendsto_at_top_is_glb h_sd (real.is_glb_Inf (set.range b) (set.range_nonempty b) h_bounded),
+end
+
+/-- The sequence `stirling_seq` is positive for `n > 0`  -/
+lemma stirling_seq'_pos (n : ℕ): 0 < stirling_seq n.succ :=
+  div_pos ((cast_pos.mpr (factorial_pos n.succ)): (0:ℝ) < _) $
+    mul_pos (real.sqrt_pos.mpr $ mul_pos two_pos $ cast_pos.mpr (succ_pos n)) $
+      pow_pos (div_pos (cast_pos.mpr (succ_pos n)) (exp_pos 1)) n.succ
```

Comment:
```suggestion
begin
  apply_rules [cast_pos.mpr, factorial_pos, exp_pos, pow_pos, div_pos, mul_pos, real.sqrt_pos.mpr,
    two_pos, succ_pos];
  apply_instance
end
```
Perhaps this is more suggestive?  I've been toying with the idea of writing a `positivity` tactic for this kind of thing, cc @robertylewis